File: games/mindustry/js/player.js
/** 玩家单位：移动、近距离手挖、随身射击 */

import { PLAYER, MAP_W, MAP_H } from './config.js';
import { getTile } from './world.js';
import { isSolidBuilding } from './buildings.js';
import { playerShoot } from './combat.js';

export function createPlayer(x, y) {
  return {
    x,
    y,
    hp: PLAYER.hp,
    maxHp: PLAYER.hp,
    speed: PLAYER.speed,
    radius: PLAYER.radius,
    mineAcc: 0,
    shootCool: 0,
    damage: PLAYER.damage,
    aim: 0,
    carrying: null, // { type } 手持一件资源时可走近核心上交
  };
}

export function tickPlayer(player, world, core, inventory, combat, input, dt) {
  // 移动
  let mx = 0;
  let my = 0;
  if (input.keys.has('KeyW') || input.keys.has('ArrowUp')) my -= 1;
  if (input.keys.has('KeyS') || input.keys.has('ArrowDown')) my += 1;
  if (input.keys.has('KeyA') || input.keys.has('ArrowLeft')) mx -= 1;
  if (input.keys.has('KeyD') || input.keys.has('ArrowRight')) mx += 1;

  // 虚拟摇杆
  if (input.stick.active) {
    mx += input.stick.x;
    my += input.stick.y;
  }

  if (mx || my) {
    const len = Math.hypot(mx, my) || 1;
    const nx = player.x + (mx / len) * player.speed * dt;
    const ny = player.y + (my / len) * player.speed * dt;
    player.x = collideMove(world, player, nx, player.y).x;
    player.y = collideMove(world, player, player.x, ny).y;
  }

  player.x = clamp(player.x, 0.3, MAP_W - 0.3);
  player.y = clamp(player.y, 0.3, MAP_H - 0.3);

  // 瞄准
  if (input.worldAim) {
    player.aim = Math.atan2(input.worldAim.y - player.y, input.worldAim.x - player.x);
  }

  // 手挖矿（按住 E / 挖矿按钮）
  if (input.mining) {
    tryMine(player, world, core, inventory, dt);
  }

  // 靠近核心自动上交手持资源
  if (player.carrying && core) {
    const cx = core.x + core.def.size / 2;
    const cy = core.y + core.def.size / 2;
    if (Math.hypot(player.x - cx, player.y - cy) < 2.2) {
      core.storage[player.carrying.type] = (core.storage[player.carrying.type] || 0) + 1;
      inventory[player.carrying.type] = (inventory[player.carrying.type] || 0) + 1;
      player.carrying = null;
    }
  }

  // 射击（空格或触屏射击）
  player.shootCool = Math.max(0, player.shootCool - dt);
  if (input.shooting && player.shootCool <= 0) {
    player.shootCool = 1 / PLAYER.fireRate;
    playerShoot(combat, player, player.aim);
  }
}

function tryMine(player, world, core, inventory, dt) {
  if (player.carrying) return;
  let best = null;
  let bestD = PLAYER.mineRange;
  const x0 = Math.floor(player.x - 2);
  const y0 = Math.floor(player.y - 2);
  const x1 = Math.floor(player.x + 2);
  const y1 = Math.floor(player.y + 2);
  for (let y = y0; y <= y1; y++) {
    for (let x = x0; x <= x1; x++) {
      const t = getTile(world, x, y);
      if (!t || !t.ore) continue;
      const d = Math.hypot(x + 0.5 - player.x, y + 0.5 - player.y);
      if (d < bestD) {
        bestD = d;
        best = t;
      }
    }
  }
  if (!best) return;
  player.mineAcc += PLAYER.mineRate * dt;
  if (player.mineAcc >= 1) {
    player.mineAcc = 0;
    // 直接进核心库存（早期上手友好）；远离核心则手持
    if (core) {
      const cx = core.x + core.def.size / 2;
      const cy = core.y + core.def.size / 2;
      if (Math.hypot(player.x - cx, player.y - cy) < 5) {
        core.storage[best.ore] = (core.storage[best.ore] || 0) + 1;
        inventory[best.ore] = (inventory[best.ore] || 0) + 1;
      } else {
        player.carrying = { type: best.ore };
      }
    }
  }
}

function collideMove(world, player, nx, ny) {
  const r = player.radius;
  // 采样四周固体
  const samples = [
    [nx - r, ny - r],
    [nx + r, ny - r],
    [nx - r, ny + r],
    [nx + r, ny + r],
    [nx, ny],
  ];
  for (const [sx, sy] of samples) {
    const t = getTile(world, Math.floor(sx), Math.floor(sy));
    if (t && isSolidBuilding(t.building)) {
      // 核心允许穿过边缘靠近
      if (t.building.type === 'core') continue;
      return { x: player.x, y: player.y };
    }
  }
  return { x: nx, y: ny };
}

function clamp(v, a, b) {
  return Math.max(a, Math.min(b, v));
}

File: steel-core/src/systems/player.rs
use bevy::prelude::*;

use crate::components::*;
use crate::config::{world_to_tile, PlayerStats, TILE};
use crate::map::{BuildMode, GameMap, GameOver, HoverTile, Inventory};

pub fn player_move(
    time: Res<Time>,
    keys: Res<ButtonInput<KeyCode>>,
    map: Res<GameMap>,
    over: Res<GameOver>,
    mut q: Query<&mut Transform, With<Player>>,
) {
    if over.done {
        return;
    }
    let Ok(mut tf) = q.get_single_mut() else {
        return;
    };

    let mut dir = Vec2::ZERO;
    if keys.pressed(KeyCode::KeyW) || keys.pressed(KeyCode::ArrowUp) {
        dir.y += 1.0;
    }
    if keys.pressed(KeyCode::KeyS) || keys.pressed(KeyCode::ArrowDown) {
        dir.y -= 1.0;
    }
    if keys.pressed(KeyCode::KeyA) || keys.pressed(KeyCode::ArrowLeft) {
        dir.x -= 1.0;
    }
    if keys.pressed(KeyCode::KeyD) || keys.pressed(KeyCode::ArrowRight) {
        dir.x += 1.0;
    }
    if dir == Vec2::ZERO {
        return;
    }
    let step = dir.normalize() * PlayerStats::SPEED * time.delta_secs();
    let next = tf.translation.truncate() + step;

    // 固体碰撞（核心可穿过边缘）
    let (tx, ty) = world_to_tile(next);
    let blocked = map.get(tx, ty).map(|t| t.solid).unwrap_or(true);
    if !blocked {
        tf.translation.x = next.x;
        tf.translation.y = next.y;
    } else {
        // 轴分离滑动
        let nx = Vec2::new(tf.translation.x + step.x, tf.translation.y);
        let (ax, ay) = world_to_tile(nx);
        if !map.get(ax, ay).map(|t| t.solid).unwrap_or(true) {
            tf.translation.x = nx.x;
        }
        let ny = Vec2::new(tf.translation.x, tf.translation.y + step.y);
        let (bx, by) = world_to_tile(ny);
        if !map.get(bx, by).map(|t| t.solid).unwrap_or(true) {
            tf.translation.y = ny.y;
        }
    }
}

pub fn player_mine(
    time: Res<Time>,
    keys: Res<ButtonInput<KeyCode>>,
    map: Res<GameMap>,
    mut inv: ResMut<Inventory>,
    over: Res<GameOver>,
    mut q: Query<(&Transform, &mut Player)>,
    core_q: Query<&Transform, With<CoreTag>>,
) {
    if over.done || !keys.pressed(KeyCode::KeyE) {
        return;
    }
    let Ok((tf, mut player)) = q.get_single_mut() else {
        return;
    };
    if player.carrying.is_some() {
        return;
    }

    let pos = tf.translation.truncate();
    let (px, py) = world_to_tile(pos);
    let mut best = None;
    let mut best_d = PlayerStats::MINE_RANGE;
    for dy in -2..=2 {
        for dx in -2..=2 {
            let x = px + dx;
            let y = py + dy;
            if let Some(ore) = map.get(x, y).and_then(|t| t.ore) {
                let wp = crate::config::tile_to_world(x, y, 1).truncate();
                let d = wp.distance(pos);
                if d < best_d {
                    best_d = d;
                    best = Some(ore);
                }
            }
        }
    }
    let Some(ore) = best else {
        return;
    };

    player.mine_acc += PlayerStats::MINE_RATE * time.delta_secs();
    if player.mine_acc >= 1.0 {
        player.mine_acc = 0.0;
        let near_core = core_q
            .iter()
            .any(|c| c.translation.truncate().distance(pos) < 5.0 * TILE);
        if near_core {
            inv.add(ore, 1);
        } else {
            player.carrying = Some(ore);
        }
    }

    // 上交
    if let Some(res) = player.carrying {
        if core_q
            .iter()
            .any(|c| c.translation.truncate().distance(pos) < 2.2 * TILE)
        {
            inv.add(res, 1);
            player.carrying = None;
        }
    }
}

pub fn player_shoot(
    time: Res<Time>,
    keys: Res<ButtonInput<KeyCode>>,
    mouse: Res<ButtonInput<MouseButton>>,
    hover: Res<HoverTile>,
    build: Res<BuildMode>,
    over: Res<GameOver>,
    mut commands: Commands,
    mut q: Query<(&Transform, &mut Player)>,
) {
    if over.done {
        return;
    }
    let Ok((tf, mut player)) = q.get_single_mut() else {
        return;
    };
    player.shoot_cd = (player.shoot_cd - time.delta_secs()).max(0.0);

    let shooting = keys.pressed(KeyCode::Space)
        || (mouse.pressed(MouseButton::Left) && build.selected.is_none());
    if !shooting || player.shoot_cd > 0.0 {
        return;
    }

    let origin = tf.translation.truncate();
    let aim = (hover.world - origin).normalize_or_zero();
    if aim == Vec2::ZERO {
        return;
    }
    player.shoot_cd = 1.0 / PlayerStats::FIRE_RATE;

    let spawn = origin + aim * 14.0;
    commands.spawn((
        Sprite {
            color: crate::config::colors::BULLET,
            custom_size: Some(Vec2::splat(6.0)),
            ..default()
        },
        Transform::from_translation(spawn.extend(6.0)),
        Bullet {
            damage: PlayerStats::DAMAGE,
            life: 0.55,
        },
        LinearVelocity(aim * 420.0),
    ));
}
